codice/ddf
# DDF-5649 G-6176 Add interface to filter warnings

#### What does this PR do?
In response to a malfunction of a query which the user runs, the user
would like a message which describes or explains the malfunction so that
the user can more easily resolve the malfunction. This PR supports
the user's aforementioned desire in that it:

 - adds an interface—SourceWarningsFilter—which allows that a Source
   which implements it filter the warnings of ProcessingDetails of a
   SourceResponse which the Source may receive from the execution of a
   query and, thereby, produce new ProcessingDetails whose warnings
   contain only each warning of the original ProcessingDetails which
   describes or explains a malfunction which occurred in the execution
   of the query

 - ensures that the implementations of SourceResponse not return null
   from getProcessingDetails(), in accordance with the dictate of the
   Javadoc of SourceResponse.getProcessingDetails()

 - refactors some of the code which relates to the changes listed above
#### Who is reviewing it? 
@mojogitoverhere 
@brianfelix 

#### Select relevant component teams: 
@codice/core-apis  
@codice/ui 

#### Ask 2 committers to review/merge the PR and tag them here.
@jrnorth
@lambeaux

#### How should this be tested?
~(will complete soon)~
~(but for real this time)~
**A full build should sufficiently test the changes made in this PR because these changes merely facilitate functionality which developers will implement in another repository which makes use of `ddf`. A test of the aforementioned functionality will occur in this other repository. These changes alter nothing about the experience which a user of DDF expects. Therefore, verification that these changes do not cause a failure in any of the tests in `ddf` suffices as insurance that the effect of these changes aligns with our expectations.**

#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: #5649 (well, this doesn't really fix the issue, but it lays the groundwork for the solution) 

#### Screenshots


#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
